crates/cubecl-runtime/src/tune/tuner.rs

```diff
@@ -47,115 +52,146 @@ impl<K: AutotuneKey> Tuner<K> {
     {
-        let operation = match self.tune_cache.try_cache(autotune_operation_set) {
+        match self.tune_cache.try_cache(autotune_operation_set.clone()) {
+            // Cache hit -> return straight away.
+            super::TuneCacheResult::Hit(ops) => return AutotuneOperation::execute(ops),
+            // Pending -> wait for the cache to be filled.
+            super::TuneCacheResult::Pending => {}
+            // Never seen before -> Start autotuning.
+            super::TuneCacheResult::Miss => {
+                self.start_autotuning(autotune_operation_set.clone(), client);
+            }
+        };
+
+        // Collect any new cache results that have come in.
+        while let Ok((key, fastest_index)) = self.results_rec.try_recv() {
+            self.tune_cache.cache_insert(key.clone(), fastest_index);
+            #[cfg(autotune_persistent_cache)]
+            {
+                let checksum = autotune_operation_set.compute_checksum();
+                self.tune_cache
+                    .persistent_cache_insert(key, checksum, fastest_index);
+                self.tune_cache.save();
+            }
+        }
+
+        // Check to see if value is now cached. If not, just use a default operation.
+        let operation = match self.tune_cache.try_cache(autotune_operation_set.clone()) {
             super::TuneCacheResult::Hit(ops) => ops,
-            super::TuneCacheResult::Miss(set) => self.autotuning(set, client),
+            _ => autotune_operation_set.fastest(0),
         };
 
-        AutotuneOperation::<Out>::execute(operation)
+        AutotuneOperation::execute(operation)
     }
 
-    fn autotuning<S, C, Out>(
+    fn start_autotuning<S, C, Out: Send + 'static>(
         &mut self,
-        autotune_operation_set: Box<dyn AutotuneOperationSet<K, Out>>,
+        autotune_operation_set: Arc<dyn AutotuneOperationSet<K, Out>>,
         client: &ComputeClient<S, C>,
-    ) -> Box<dyn AutotuneOperation<Out>>
-    where
-        S: ComputeServer,
-        C: ComputeChannel<S>,
+    ) where
+        S: ComputeServer + 'static,
+        C: ComputeChannel<S> + 'static,
     {
         let key = autotune_operation_set.key();
-        let autotunables = autotune_operation_set.autotunables();
-        let mut names = Vec::with_capacity(autotunables.len());
-
-        let results: Vec<Result<BenchmarkDurations, BenchError>> = autotunables
-            .into_iter()
-            .enumerate()
-            .map(|(i, op)| {
-                names.push(op.name().to_string());
-                if autotune_operation_set.should_run(&key, i) {
-                    self.run_benchmark(op, client)
-                } else {
-                    Ok(BenchmarkDurations::new(Vec::from([Duration::MAX])))
+        self.tune_cache.mark_pending(key.clone());
+
+        let set = autotune_operation_set.clone();
+        let client = client.clone();
+        let sender = self.results_send.clone();
+
+        let fut = async move {
```

Comment:
I think the async block could become another method, it's starting to become a long 😅

Reply:
Hm the code before the async block isn't doing much except cloning for the captures in the block. I've tried instead to shorten the async block a bit & move out some stuff in it.